docs: market params design_doc (#1152)
umee-network/umee: docs/design_docs/010-market-params.md

# ADR 010: Market Parameters

## Changelog

- Jul 4, 2022: Max collateral utilization added to ADR-003 (@robert-zaremba)

> - Jul 18, 2022: Initial draft moved discussion to ADR-101 (@toteki)

Reviewer: ```suggestion
- Jul 18, 2022: Initial draft moved discussion to ADR-010 (@toteki)
```

Author: Note: Fixed in document moving/renaming PR

### Maximum Collateral Share

Another proposed parameter is `MaxCollateralShare`, to be defined per token.

```go
total_collateral_value(token) = total_collateral(token) * oracle_price_usd(token)
collateral_share(token) = total_collateral_value(token) / Sum_All_Tokens(total_collateral_value) // ranges 0 - 1
```

For example, if `10 ATOM` collateral exists at a price of `$20` alongside `100 OSMO` collateral at a price of `$1`, then `collateral_share(ATOM)` is `10 * $20 / $300 = 0.66` and `collateral_share(OSMO)` is `$100 / $300 = 0.33`.

Implementing `MaxCollateralShare` would restrict `MsgCollateralize` from increasing `total_collateral` above a desired level.

`MaxCollateralShare` could still be indirectly exceeded by fluctuating oracle prices, `MsgDecollateralize` or `MsgLiquidate` of other tokens, or interest accruing in one denom faster than another.

### Minimum Collateral Liquidity

Another proposed parameter is `MinCollateralLiquidity`, to be defined per token.

It is the reciprocal of `MaxCollateralUtilization`, functionally equivalent in which messages it would restrict.

```go

> collateral_liquidity(token) = available(token) / total_collateral(token) // ranges 0 - ∞

Reviewer: ```suggestion
collateral_liquidity(token) = available(token) / total_collateral(token) // [0, ∞]
```

Implementing `MinCollateralLiquidity` would restrict `MsgBorrow` and `MsgWithdraw` from decreasing `available_supply`, or `MsgCollateralize` from increasing `total_collateral` in certain market conditions.

It may or may not allow `MsgLiquidate` to decrease `available_supply` under the same conditions, to prevent crises.

| Message Type | Current Decision |
| - | - |
| `MsgBorrow` | Restrict |
| `MsgWithdraw` | Restrict |
| `MsgLiquidate` | Allow |

`MinCollateralLiquidity` could still be indirectly exceeded by supply interest accruing on a collateral denom's uToken exchange rate.

It has also been proposed separately to factor `collateral_liquidity` (or `collateral_utilization`) into dynamic interest rates, to enhance the current model which uses `supply_utilization` (see ADR-004).

## Alternatives

### Collateral Utilization

One proposed parameter is `MaxCollateralUtilization`, to be defined per token.

It is the reciprocal of `MinCollateralLiquidity`, so the motivation and tradeoffs will be the same.

```go

> collateral_utilization(token) = total_collateral(token) / available(token) // ranges 0 - ∞

Reviewer: ```suggestion
collateral_utilization(token) = total_collateral(token) / available(token) // [0 - ∞]
```

```

For example if `200 ATOM` of collateral exists and the `x/leverage` module only has `40 ATOM` available (with the rest of supply being borrowed out), then `collateral_utilization(ATOM) = 5`.

This quantity has the property of increasing rapidly (as 1/N -> 0) when available supply is under stress.

> Stable assets will have high max collateral utilization (can go even to 50). Volatile assets should have significantly smaller collateral utilization, and assets with high risk should have max collateral utilization close to 1.

Reviewer: ```suggestion
Stable assets will have high max collateral utilization (up to to 50). Volatile assets should have significantly smaller collateral utilization, and assets with high risk should have max collateral utilization close to 1.
```
